llvm/lib/Transforms/Yk/ShadowStack.cpp

```diff
@@ -0,0 +1,230 @@
+//===- ShadowStack.cpp - Pass to add shadow stacks to the AOT module --===//
+//
+// Add shadow stacks to store variables that may have their references taken.
+// Storing such variables on a shadow stack allows AOT to share them with
+// compiled traces, and back (i.e. references created inside a trace will still
+// be valid when we return from the trace via deoptimisation).
+// YKFIXME: This can be optimised by only putting variables on the shadow stack
+// that actually have their reference taken.
+
+#include "llvm/Transforms/Yk/ShadowStack.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Transforms/Yk/LivenessAnalysis.h"
+
+#include <map>
+
+#define DEBUG_TYPE "yk-shadowstack"
+// The size of the shadow stack. Defaults to 1MB.
+// YKFIXME: Make this adjustable by a compiler flag.
+#define SHADOW_STACK_SIZE 1000000
+
+using namespace llvm;
+
+namespace llvm {
+void initializeYkShadowStackPass(PassRegistry &);
+} // namespace llvm
+
+namespace {
+class YkShadowStack : public ModulePass {
+public:
+  static char ID;
+  YkShadowStack() : ModulePass(ID) {
+    initializeYkShadowStackPass(*PassRegistry::getPassRegistry());
+  }
+
+  // Checks whether the given instruction is the alloca of the call to
+  // `yk_mt_new`.
+  bool isYkMTNewAlloca(Instruction *I) {
+    for (User *U: I->users()) {
+      if(U && isa<StoreInst>(U)) {
+        Value *V = cast<StoreInst>(U)->getValueOperand();
+        if(isa<CallInst>(V)) {
+          CallInst *CI = cast<CallInst>(V);
+          if (CI->isInlineAsm())
+            return false;
+          if (!CI->getCalledFunction())
+            return false;
+          return (CI->getCalledFunction()->getName() == "yk_mt_new");
```

Comment:
`yk_mt_new` should probably be a constant (I *think* it's a constant elsewhere? it would be good to have one constant for ykllvm as a whole, if possible).

Reply:
Fixed in [a0b9ada](https://github.com/ykjit/ykllvm/pull/53/commits/a0b9ada8ca16e730f8cc91909ef144e718be7fc9)